netbox/client.go

```diff
@@ -0,0 +1,173 @@
+// header) and a map[string]string of Endpoints
+type API struct {
+	Version   string
+	Endpoints map[string]string
+}
+
+// NewClient returns a new instance of a NetBox client.  addr specifies the address
+// of the NetBox server, and client specifies an optional HTTP client to use
+// for requests.
+//
+// If client is nil, a default HTTP client will be used.
+func NewClient(addr string, client *http.Client) (*Client, error) {
+	if client == nil {
+		client = &http.Client{}
+	}
+
+	u, err := url.Parse(addr)
+	if err != nil {
+		return nil, err
+	}
+
+	c := &Client{
+		u:      u,
+		client: client,
+	}
+
+	c.DCIM = &DCIMService{c: c}
+	c.IPAM = &IPAMService{c: c}
+
+	return c, nil
+}
+
+// NewRequest creates a HTTP request using the input HTTP method, URL
+// endpoint, and a Valuer which creates URL parameters for the request.
+//
+// If a nil Valuer is specified, no query parameters will be sent with the
+// request.
+func (c *Client) NewRequest(method string, endpoint string, options Valuer) (*http.Request, error) {
+	rel, err := url.Parse(endpoint)
+	if err != nil {
+		return nil, err
+	}
+
+	// Allow specifying a base path for API requests, so if a NetBox server
+	// resides at a path like http://example.com/netbox/, API requests will
+	// be sent to http://example.com/netbox/api/...
+	//
+	// Enables support of: https://github.com/digitalocean/netbox/issues/212.
+	if c.u.Path != "" {
+		rel.Path = path.Join(c.u.Path, rel.Path)
+	}
+
+	u := c.u.ResolveReference(rel)
+
+	// If no valuer specified, create a request with no query parameters
+	if options == nil {
+		return http.NewRequest(method, u.String(), nil)
+	}
+
+	values, err := options.values()
+	if err != nil {
+		return nil, err
+	}
+	u.RawQuery = values.Encode()
+
+	return http.NewRequest(method, u.String(), nil)
+}
+
+// Do executes an HTTP request and if v is not nil, Do unmarshals result
+// JSON onto v.
+func (c *Client) Do(req *http.Request, v interface{}) error {
+	res, err := c.client.Do(req)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		_ = res.Body.Close()
+	}()
+	// Test if the request was successful.
+	// If not, do not try to decode the body. This would result in
+	// misleading error messages
+	err = httpStatusOK(res)
+	if err != nil {
+		return err
+	}
+
+	if v == nil {
+		return nil
+	}
+
+	return json.NewDecoder(res.Body).Decode(v)
+}
+
+// httpStatusOK tests if the StatusCode of res is smaller than 300
+// It returns an error (or nil)
+func httpStatusOK(res *http.Response) error {
+	if res.StatusCode >= http.StatusMultipleChoices {
+		bodyData, err := ioutil.ReadAll(res.Body)
+		if err != nil {
+			return fmt.Errorf("%d - %v", res.StatusCode, err)
+		}
+		return fmt.Errorf("%d - %s", res.StatusCode, bodyData)
+	}
+	return nil
+}
+
+// Ping sends a get request to /api/. It Returns an API struct, containing the
+// API-Version (extracted from the response header) and the list of endpoints
+// as a map[string]string
+func (c *Client) Ping() (*API, error) {
```

Comment:
I though of it as a way to check the connection, directly after NewCient. When I started to use go-netbox-v1, I kind of struggled because of an unauthorized error which was not cleanly given back to my implementation.

So in fact, we don't need this. Since I changed the Do-implementation, to somehow handle != 2xx status codes, I could toss Ping. What do you think?